### PR TITLE
internal/dag: Look for duplicate header conditions and mark Proxy as invalid

### LINKED ISSUE #1559: internal/dag: Look for duplicate headers in a route configuration
**Describe the solution you'd like**
An HTTPProx's route can end up with many Conditions applied to it through various levels of Includes. 

Contour should verify that the resulting set of Conditions do not overlap. This is specific for headers. If any set of headers duplicate key's, then an error should be set on the Proxy which is malformed.

### PR DESCRIPTION
**Pull request by stevesloka**
Fixes #1559 by looking for any route that has duplicate header keys and marking them as invalid if found. The logic this includes is to only mark duplicate headers with the "exact" match type. I could see folks setting up multiple "notexact" or "notcontains" headers which should be logical match types.

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: internal/dag/builder.go [hunk 1]
```diff
@@ -518,6 +518,13 @@ func (b *Builder) computeRoutes(sw *ObjectStatusWriter, proxy *projcontour.HTTPP
 			}
 
 			sw, commit := b.WithObject(delegate)
+
+			//	conds := append(conditions, include.Conditions...)
+			//if !headerConditionsAreValid(conds) {
+			//	sw.SetInvalid("cannot specify duplicate header conditions in the same include")
+			//	return nil
+			//}
+
```
Comment (davecheney) [inline internal/dag/builder.go]: please don't commit commented out code.

File: internal/dag/conditions.go [hunk 2]
```diff
@@ -70,3 +70,22 @@ func headerConditions(conds []projcontour.Condition) []HeaderCondition {
...
+	// Look for duplicate headers on conditions
+	// if found, set error condition on HTTPProxy
+	encountered := map[string]bool{}
+	for _, v := range conditions {
+		if v.Header == nil {
+			continue
+		}
+		switch {
+		case v.Header.Exact != "":
+			if encountered[v.Header.Name] {
```
Comment (davecheney) [inline internal/dag/conditions.go]: I think this needs to be a case insensitive comparison; this test case https://github.com/projectcontour/contour/blob/master/internal/dag/conditions_test.go#L196 should fail.

**Comment by stevesloka** [comment]
> TL;DR I'm not sure we should throw out duplicate header matching conditions. I think there is a legitimate use case, you can easily say header "foo:" contains both "a" and a later conditions is contains "b". We should agree on the business logic before duking it out in test cases.

This code only throws out duplicate headers if they are of type `exact` match. My thought was that if you specified header `foo: a` should match exactly, then later also said that it should match `foo: b` you'd essentially create a scenario where a route won't get requests. 

Every other type of header is allowed. For example, I could say that a request does not contain "foo: a" as well as should not contain "foo: b".

**Comment by davecheney** [comment]
Ahh ok. I’m sorry that was my mistake, I didn’t read the code clearly enough.

**Comment by stevesloka** [comment]
Let me clean up the code a bit to make that more clear.
